### PR TITLE
[ ZEPPELIN-967 ] Implement search bar in job manager

### PR DESCRIPTION
**Pull request by cloverhearts**
### What is this PR for?

A few sentences describing the overall goals of the pull request's commits.
First time? Check out the contributing guide - https://zeppelin.apache.org/contribution/contributions.html
### What type of PR is it?

Feature
### Todos
- [x] add serach bar ui
- [x] implement filter
- [x] show filter loading status.
### What is the Jira issue?

https://issues.apache.org/jira/browse/ZEPPELIN-967
### How should this be tested?
1. click to jobmanager menu
2. try to use filter by jobmanager filter bar on top.

Enter, Enter, it has been filtered properly?
After inputting, wait about 2 seconds, it has been filtering?
### Screenshots (if appropriate)

<a href="https://cloud.githubusercontent.com/assets/10525473/18658644/d6359a62-7f3f-11e6-8c03-030fc107ec2d.gif">Screenshots</a>

https://cloud.githubusercontent.com/assets/10060731/20595003/a69386ce-b27b-11e6-936f-dc6f78db7e94.gif

### Questions:
- Does the licenses files need update? no
- Is there breaking changes for older versions? no
- Does this needs documentation? no

### CODE REVIEW

**Comment by corneadoug** [comment]
@cloverhearts Thanks for the contribution, it looks cool.
I will take a look :)

**Comment by cloverhearts** [comment]
@corneadoug Thank you!

**Comment by cloverhearts** [comment]
ping

**Comment by cloverhearts** [comment]
code rebase

**Comment by cloverhearts** [comment]
@corneadoug 
sorry, could you please review for this pr?

**Comment by AhyoungRyu** [comment]
Hi @cloverhearts. Thanks for the improvement!
I just tested and got some small feedbacks :)

1.  Is there any particular reason to have dropdown menu next to the filter bar? If so, seems need to be sorted to alphabetic order or something 
screen shot 2016-10-06 at 2 43 29 pm

2. As you can see in the below gif image, there is latency after some typing. 
latency
We have a good example in Interpreter menu :) The search bar in interpreter menu, it shows the searched result right after typing. 

3. The applied font in dropdown menu and the filter bar is different. 
screen shot 2016-10-06 at 2 56 05 pm

**Comment by corneadoug** [comment]
@cloverhearts Just tested and checked the code.
Overall same feedbacks as @AhyoungRyu 

#### Filtering system
The one you used seems a bit too complicated (also the filter should have been in a new file).
Since it is also slow, using the same method used in interpreter search would be better.

#### Filtering input
Try to reuse the same as the interpreter search, it should have the proper font and keep the same style between pages.

#### Interpreter dropdown filtering
I would remove this feature mainly because it only works with the default interpreter.
Maybe also remove the default interpreter written on each job?

#### No results
Current message is: `Data does not exist`, `No Job found` would be better

#### Possible future work that could use a Jira issue
* Removing the legend (ready, finished, abort etc...), that info is already on each dot.
* Make Interpreter and Job page top part fixed, so that we can search even while scrolling

**Comment by cloverhearts** [comment]
@AhyoungRyu @corneadoug 
Thank you very much for your feedback.

#### Filtering system
I think the filter I added here actually will improve filtering performance since as soon as you press return key, it will show filtering result. And if there is no input from user for more than 2 sec, it stops filtering.

The reason I implemented filtering myself is to fix the latency issue with large number of notebooks. If you filter on every keydown event, it will freeze browser when you have a lot of notebooks.(tested with 250 notebooks). 

Will it be better if I add UI letting user know that they should press return button to complete filtering?


@corneadoug 
actually, I've used the interpreter filtering in the first implementation.
However, the behavior is a bit different from job filtering. So I had to define separate filters.

#### Filtering input and No results
Your opinion is right.
Okay, I will fix it.

#### Interpreter dropdown filtering
I think this feature is very important.
because currently there is no way to manage your interpreter which are not bound to notebook.
also, if you use a lot of notebooks and interpreter, I think users would want to see and sort it.
Personally I also wanted to have this kind of view when i was using cron scheduler to operating a notebook then can not find problem for interpreter bind.
I think this feature is a good way to be feedback to the user.

#### Possible future work that could use a jira issue
Yes, I agree with your opinion.
i will make a jira issue and implement those in future.

Thank you again.

**Comment by cloverhearts** [comment]
replace

todo list
- [x] Create jobmanager.filter.js (divide file for filter logic)
- [x] Change the filter of the font to the Zeppelin.
- [x] remove legend.

**Comment by cloverhearts** [comment]
rebase done.

**Comment by cloverhearts** [comment]
imporove

add Search button.
If you enter the return key or click the search button, the search function will work.

**Comment by cloverhearts** [comment]
CI Error

This is irrelevant to my edits.
It is also verified in other PRs.

```
19:24:48,843  INFO org.apache.zeppelin.integration.ZeppelinIT:189 - testCreateNotebook Test executed
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 146.506 sec - in org.apache.zeppelin.integration.ZeppelinIT

Results :

Tests in error: 
  ParagraphActionsIT.testEditOnDoubleClick:443 » ElementNotVisible Element is no...

Tests run: 18, Failures: 0, Errors: 1, Skipped: 0

```

**Comment by cloverhearts** [comment]
@corneadoug @AhyoungRyu 
Hello,
Could you please review this for pr?
I think the merger is ready.
Thank you

hava a nice day :)

**Comment by bzz** [comment]
@cloverhearts looks like CI is failing. Do you know the reason? If you try open\close the PR, can it be reproduced?

**Comment by cloverhearts** [comment]
@bzz 
Thank you for your feedback.
The problem with CI is irrelevant to my own.
I think this is probably related to this.

Https://github.com/apache/zeppelin/commit/4dd36bfa312285a13347fb15beadcb0baa30ed23

I will rebase first, then I'll watch the CI situation a little bit more.

Thank you!

```
19:24:48,843  INFO org.apache.zeppelin.integration.ZeppelinIT:189 - testCreateNotebook Test executed
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 146.506 sec - in org.apache.zeppelin.integration.ZeppelinIT

Results :

Tests in error: 
  ParagraphActionsIT.testEditOnDoubleClick:443 » ElementNotVisible Element is no...

Tests run: 18, Failures: 0, Errors: 1, Skipped: 0

...



```

**Comment by cloverhearts** [comment]
CI Green!

**Comment by AhyoungRyu** [comment]
@cloverhearts Nice. Glad to see the green light at last! 
Sorry for my late response. Let me take a look again and will ping you :)

**Comment by cloverhearts** [comment]
@AhyoungRyu 
Thank you very much!

**Comment by AhyoungRyu** [comment]
@cloverhearts 
Tested again and it works well as expected. Seems there are some changes after I saw this 2 month ago. 
 - The result is reset after 10 sec
 - User should click search button (or press enter) to run query. Yeah I know auto result printing is not good in performance point of view. 

I just want to make sure :)
And I left a comment about really minor UI issue. Except that it looks good to me 👍 

(p.s. It'll be great if you can update PR description. Then other ppl who are interested in this PR can look through quickly!)

File: zeppelin-web/src/app/jobmanager/jobmanager.html [hunk 1]
```diff
@@ -32,8 +32,42 @@ <h3 class="new_h3">
...
 
   <div class="row">
+    <div class="col-md-6 text-left">
+      <div class="form-inline">
+        <span class="labelBtn btn-group" style="margin-left: 0px;">
+          <button type="button" class="btn btn-default btn-xs dropdown-toggle"
+                 data-toggle="dropdown"
+                 style="width:100px; text-align: right !important;">
+           <span class="text-right job-note-name-font-family">
+             {{filterValueToName(filterConfig.filterValueInterpreter)}}<span class="caret"></span>
```
Comment (cloverhearts) [inline zeppelin-web/src/app/jobmanager/jobmanager.html]: @AhyoungRyu 
Okay, that's simple!

**Comment by cloverhearts** [comment]
@AhyoungRyu 
newnew


fix done.
Thank you :)

**Comment by AhyoungRyu** [comment]
@cloverhearts Thanks for addressing my comment in here. 
Seems CI is happy now and the feature is working well i think. But since I'm not a front-end expert, it would be better if other ppl can review this one before merging into master.

**Comment by cloverhearts** [comment]
@Leemoonsoo 
Thank you for your review a lot.

First, the select box's interpreter shows only the first interpreter, because it is information about the default interpreter's settings.
In fact, the meaning of this information is that it makes sense to find an `uninterpreted note '.

your said pointed all the interpreters, I will implement on this pr later.

The navigation bar of the existing navbar and the search menu of the `job task` menu have different meanings.

actually, navbar's search box is a good way to view detailed information about all the para-graphs.

but, I think we need search feature on `job menu(task)` are real-time response task monitoring.
the search for the `job task` should now filter the real-time information.

For example, if multiple users create and run a new paragraph or note, the search results of the `job task` should be changed to match the current situation.
The search using lucene is the information at the time the search query was requested.
It is detailed but not suitable for `job monitoring`.

I also think that the information on the `job task` can contain more information about the current work situation.
I have made a little more preparation for this part. (also your said point too)
That's why I think the search for the current navbar's search box and the `job task` should go to a different implementation.

**Comment by cloverhearts** [comment]
change name for item on select box.
Undefined to Interpreter is not set

**Comment by Leemoonsoo** [comment]
@cloverhearts Thanks for the explanation. Keeping search with real time job monitoring make sense.

LGTM and merge to master. if there're no further discussions.

**Pull request closed without merging**